### ISSUE TITLE
[API]: ForwardInterceptor overwrites existing CustomRecords

### ISSUE THREAD

**Issue by carlaKC** [lightningnetwork/lnd#9166]
This is an API design question, not a bug.

We're interested in the following scenario:
1) The forward interceptor is used with the `MODIFIED` action
2) We're intercepting a HTLC with a custom range endorsed TLV
3) The interceptor request's [out_wire_custom_records](https://github.com/lightningnetwork/lnd/blob/8e451cb8be86267864e15c253e2b174849f4e70f/lnrpc/routerrpc/router.proto#L1066) are either:
  3.1)  Populated by the interceptor request
  3.2) Empty

I'm using the endorsement field in (2) because I hit this in #8390, but the same would apply to any custom record that LND may want to attach by default in the future. If we don't forsee LND _ever_ appending custom records to HTLCs then this isn't as interesting of a question. 

We have different API behaviors depending on whether or not `out_wire_custom_records` are set by the interceptor:
3.1) If the interceptor sets custom records, we only propagate the records provided in the interceptor request
3.2) If the interceptor request does not provide custom records, we propagate whatever was already there

This happens because we [overwrite](https://github.com/lightningnetwork/lnd/blob/8e451cb8be86267864e15c253e2b174849f4e70f/htlcswitch/interceptable_switch.go#L719) `htlc.CustomRecords` if the interceptor sets a value, and otherwise leave as-is.

I think it would be more consistent for the interceptor to _always_ require that the `MODIFIED` request provides all of the TLVs they want to propagate - ie, do not passively propagate the records in the 3.2 case. This has the downside of requiring the caller to copy all custom records that they're _not_ concerned with from response to request, but I think is cleaner than the alternative of trying to specify mutation of the existing records via protobuf.

**Comment by ProofOfKeags** [comment]
I think the ideal API here allows for two cases:
1. don't touch it: leave it completely unmodified, I don't even want to have to think about it
2. full replacement: I am taking responsibility for the records, give the incoming ones as input, I will do a full replacement

My intuition here is that we should make an empty map be replacement with no records, and an *absent* map behave as the identity function (leave unmodified). However, this is where the details of protobuf encodings might get in our way. I believe that `repeated` in protobuf means that things can appear zero times. This would mean that there is no way to encode a *present-but-empty* map. Assuming we can't make this distinction, I think it's probably best that we opt for full-replacement even though that may require user action.

By inserting an interceptor into the workflow the user is claiming responsibility for interceding an otherwise automatic process. Losing the ability to delete records as outlined in #9215 is (in my view) not a good tradeoff here. It makes the API confusing and unpredictable (which accumulates and compounds) just to make things slightly easier for a process where the user has already claimed the responsibility.

If this suggestion is untenable, I think we should go with a full-merge based approach by:
1. having an output for deleted keys
2. having an output for added (replace on conflict) key-value pairs
We always apply 1 before 2 here as I believe the sensible interpretation of having included them in both deleted and added means that they definitely wanted the stuff they added, not just to delete it.

That said, I do believe that this will be less intuitive and easier to get wrong as a user. Full-replacement is a more easily explained set of semantics.

**Comment by carlaKC** [comment]
> This would mean that there is no way to encode a present-but-empty map. Assuming we can't make this distinction

For [other APIs](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L2280) we've gone with a bool to distinguish between setting a zero value and no value at all, that's also an option here:
* Always merge the record map with existing fields (replace on conflict)
* Add a `clear_custom_records` option to clear `custom_records` (which must be set alongside an empty map)

This came up on the last contributor call and it seemed to me like there wasn't much enthusiasm to make any changes here, so the questions isn't really about API design - it's whether this is something that'll be written/reviewed/merged. I'm happy to review a change fwiw, but don't have a compelling enough use case/opinion to write the PR.

**Comment by ProofOfKeags** [comment]
> I'm happy to review a change fwiw, but don't have a compelling enough use case/opinion to write the PR.

Seems reasonable.

> Add a clear_custom_records option to clear custom_records (which must be set alongside an empty map)

If we go this route, then assuming we don't actually implement this yet, but we want the semantics to resemble the above AND we want the future change to be backwards compatible. Then that means that we would *start* with a full-merge API. Which iirc is not the way things are designed right?

**Comment by carlaKC** [comment]
> Then that means that we would start with a full-merge API. Which iirc is not the way things are designed right?

Yes that's correct. Seems like a reasonable (and small) to update the API to merge (/overwrite) when records are populated on the request - @guggero would that work for your current use case?

**Comment by guggero** [comment]
Sure, I think that should work for our use case. I think it's just important to make those changes before we ship this new functionality with a tagged release. So the question is: do you have time to do the update? And would it be a new PR on top of the documentation change? Ideally we'd merge the doc change PR, then fix it with the new behavior in a new PR, then the diff to today's functionality will be even more clear.

**Comment by carlaKC** [comment]
>  Ideally we'd merge the doc change PR, then fix it with the new behavior in a new PR, then the diff to today's functionality will be even more clear.

SGTM! Let's go ahead with merging #9215 and I'll follow with the fix, looks [pretty simple](https://github.com/carlaKC/lnd/commit/21316031d9452dbb5ee76abe2e1ce546348ff7db). 

It is a bit difficult to test, so I think I'll rebase #8390 on it so that we can test the case where LND already has some records?
